src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -57,11 +97,82 @@ private void AddExceptionDiagnostic(Diagnostic diagnostic)
-            return compilerDiagnostics.AddRange(analyzerDiagnostics).AddRange(_exceptionDiagnostics);
+                lock (_analysisLock)
+                {
+                    if (!_documentTasks.TryGetValue(documentTree, out documentTask))
+                    {
+                        if (_completeAnalysisTask != null)
+                        {
+                            // Once complete analysis has begun, it is too late to start analysis of an individual document.
+                            return ImmutableArray<Diagnostic>.Empty;
+                        }
+
+                        if (latestAnalysisTask != _latestAnalysisTask)
+                        {
+                            // Another analysis task has started. Wait to start this one.
+                            // Analysis tasks are serialized because the driver can use all available cores
+                            // in one analysis and swamping the system with tasks is not advantageous.
+                            continue;
+                        }
+
+                        _analyzedSyntaxTrees = _analyzedSyntaxTrees.Add(documentTree);
+
+                        documentTask = Task.Run(async () =>
+                        {
+                            // Invoke GetDiagnostics to populate the compilation's CompilationEvent queue.
+                            // Discard the returned diagnostics.
+                            model.GetDiagnostics(null, _cancellationToken);
+
+                            ImmutableArray<Diagnostic> analyzerDiagnostics = await _driver.GetPartialDiagnosticsAsync(documentTree, _cancellationToken).ConfigureAwait(false);
+                            return analyzerDiagnostics;
+                        }, _cancellationToken);
+
+                        _documentTasks[documentTree] = documentTask;
+                        _latestAnalysisTask = documentTask;
+                    }
+                }
+
+                return await documentTask.ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Returns diagnostics produced compilation and by diagnostic analyzers from analyzing a single document.
+        /// <param name="model">Semantic model for the document.</param>
+        /// </summary>
+        public async Task<ImmutableArray<Diagnostic>> GetDiagnosticsFromDocumentAsync(SemanticModel model)
+        {
+            ImmutableArray<Diagnostic> analyzerDiagnostics = await GetAnalyzerDiagnosticsFromDocumentAsync(model).ConfigureAwait(false);
+            return model.GetDiagnostics(null, _cancellationToken).AddRange(analyzerDiagnostics);
+        }
+        
+        /// <summary>
+        /// Returns diagnostics produced for exceptions thrown by analyzer actions.
+        /// </summary>
+        public ImmutableArray<Diagnostic> GetExceptionDiagnostics()
+        {
+            return _exceptionDiagnostics.ToImmutableArrayOrEmpty();
```

Comment:
Will this return the correct value if a non-null onAnalyzerException delegate was provided? We may want to ensure that `_exceptionDiagnostics` is always populated with exception diagnostics, regardless of whether onAnalyzerException is null or not.